### PR TITLE
Refactor to Puppet 4 types

### PR DESCRIPTION
**Pull request by ekohl**

### CODE REVIEW

**Comment by ekohl** [comment]
This is just a WIP for now, just to see how it goes.

**Comment by ekohl** [comment]
I did a Katello install with it and it appeared to work. Careful review is needed as well as https://github.com/Katello/foreman-installer-modulesync/pull/25

File: manifests/init.pp [hunk 1]
```diff
@@ -283,171 +232,123 @@
...
-  $disabled_authenticators   = $pulp::params::disabled_authenticators,
-  $additional_wsgi_scripts   = $pulp::params::additional_wsgi_scripts,
-  $puppet_wsgi_processes     = $pulp::params::puppet_wsgi_processes,
-  $migrate_db_timeout        = $pulp::params::migrate_db_timeout,
-  $show_conf_diff            = $pulp::params::show_conf_diff,
-  $enable_profiling          = $pulp::params::enable_profiling,
-  $profiling_directory       = $pulp::params::profiling_directory,
+  String $version = $::pulp::params::version,
+  Boolean $crane_debug = $::pulp::params::crane_debug,
+  Integer[0, 65535] $crane_port = $::pulp::params::crane_port,
```
Comment (stbenjam) [inline manifests/init.pp]: We use this so often, would it make sense to make a Port type, maybe in puppet-stdlib?

File: manifests/init.pp [hunk 2]
```diff
@@ -283,171 +232,123 @@
...
+  Optional[Stdlib::Absolutepath] $messaging_client_cert = $::pulp::params::messaging_client_cert,
+  String $messaging_topic_exchange = $::pulp::params::messaging_topic_exchange,
+  Boolean $messaging_event_notifications_enabled = $::pulp::params::messaging_event_notifications_enabled,
+  Optional[String] $messaging_event_notification_url = $::pulp::params::messaging_event_notification_url,
+  String $messaging_version = $::pulp::params::messaging_version,
+  String $broker_url = $::pulp::params::broker_url,
+  Boolean $broker_use_ssl = $::pulp::params::broker_use_ssl,
+  Optional[String] $tasks_login_method = $::pulp::params::tasks_login_method,
+  String $email_host = $::pulp::params::email_host,
+  Integer $email_port = $::pulp::params::email_port,
```
Comment (stbenjam) [inline manifests/init.pp]: `Integer[0, 65535]`

File: manifests/init.pp [hunk 3]
```diff
@@ -283,171 +232,123 @@
...
+  String $email_from = $::pulp::params::email_from,
+  Boolean $email_enabled = $::pulp::params::email_enabled,
+  Boolean $manage_squid = $::pulp::params::manage_squid,
+  Optional[String] $lazy_redirect_host = $::pulp::params::lazy_redirect_host,
+  Optional[Integer[0, 65535]] $lazy_redirect_port = $::pulp::params::lazy_redirect_port,
```
Comment (stbenjam) [inline manifests/init.pp]: Nitpick, technically ports you *connect to* should start at 1, IIRC `0` is only an indication when listening that you'll take any free port.

Comment (stbenjam) [inline manifests/init.pp]: (Not really sure we care about that differentiation)

File: manifests/init.pp [hunk 4]
```diff
@@ -283,171 +232,123 @@
...
+  Boolean $repo_auth = $::pulp::params::repo_auth,
+  Optional[String] $proxy_url = $::pulp::params::proxy_url,
+  Optional[Integer[0, 65535]] $proxy_port = $::pulp::params::proxy_port,
+  Optional[String] $proxy_username = $::pulp::params::proxy_username,
+  Optional[String] $proxy_password = $::pulp::params::proxy_password,
+  Optional[String] $yum_max_speed = $::pulp::params::yum_max_speed,
+  Integer[0] $num_workers = $::pulp::params::num_workers,
+  Boolean $enable_katello = $::pulp::params::enable_katello,
+  Boolean $enable_crane = $::pulp::params::enable_crane,
+  Optional[Integer[0]] $max_tasks_per_child = $::pulp::params::max_tasks_per_child,
```
Comment (stbenjam) [inline manifests/init.pp]: This variable is actually missing from params.pp

File: manifests/init.pp [hunk 5]
```diff
@@ -283,171 +232,123 @@
...
+  Optional[String] $lazy_redirect_host = $::pulp::params::lazy_redirect_host,
+  Optional[Integer[0, 65535]] $lazy_redirect_port = $::pulp::params::lazy_redirect_port,
+  Optional[String] $lazy_redirect_path = $::pulp::params::lazy_redirect_path,
+  Boolean $lazy_https_retrieval = $::pulp::params::lazy_https_retrieval,
+  Integer[0] $lazy_download_interval = $::pulp::params::lazy_download_interval,
+  Integer[0] $lazy_download_concurrency = $::pulp::params::lazy_download_concurrency,
+  Optional[Stdlib::Absolutepath] $consumers_crl = $::pulp::params::consumers_crl,
+  Boolean $reset_cache = $::pulp::params::reset_cache,
+  Enum['none', 'optional', 'require', 'optional_no_ca'] $ssl_verify_client = $::pulp::params::ssl_verify_client,
+  Variant[Array[String], String] $ssl_protocol = $::pulp::params::ssl_protocol,
```
Comment (stbenjam) [inline manifests/init.pp]: 👍

**Comment by stbenjam** [comment]
Generally looks good to me 👍

**Comment by stbenjam** [comment]
Incidentally, is there any script to automate this conversion to Puppet 4 types? I wrote a quick bit of hideous ruby that'll do it https://gist.github.com/stbenjam/cb602979479ed4466d83c175f0f63036, and ran in Katello https://github.com/Katello/puppet-katello/pull/188

**Comment by ekohl** [comment]
I'm on mobile so can't easily link but I know Dominic had a gist that uses the kafo parsers for this. It works well

**Comment by ekohl** [comment]
The gist from Dominic is https://gist.github.com/domcleal/5c1c9e892d175ac351d894204e7069db

**Comment by stbenjam** [comment]
Ah I see thanks.  Mine actually edits the .pp in place, albeit it's uglier and regexier.

File: manifests/init.pp (hunk 3 above)
Comment (ekohl) [inline manifests/init.pp]: You're right, I'll change that.

File: manifests/init.pp (hunk 2 above)
Comment (ekohl) [inline manifests/init.pp]: Done

File: manifests/init.pp (hunk 1 above)
Comment (ekohl) [inline manifests/init.pp]: I'll see about that, but let's start with these static types.

File: manifests/init.pp (hunk 4 above)
Comment (ekohl) [inline manifests/init.pp]: https://github.com/Katello/puppet-pulp/blob/74fb5e2c9b666253ae0bc3b4465f2344b8f235d2/manifests/params.pp#L111 is there, right?

**Comment by ekohl** [comment]
Updated and accidentally included https://github.com/Katello/puppet-pulp/pull/229. Won't rebase till the issue in stdlib is fixed to save some resources.

**Comment by jlsherrill** [comment]
ACK thanks @ekohl !!!

**Pull request merged**